### ISSUE TITLE
Fix what happens when every unit is eventually treated

### ISSUE THREAD

**Issue by kylebutts** [kylebutts/did2s_stata#12]
Currently, the first_stage regression will produce point estimates in the periods after everyone is treated (assuming those fixed effects are zero). This is not correct and should instead drop those years all together and use the last treated group as the "control group"

**Comment by kylebutts** [comment]
This is the problem distilled:

``` stata
. sysuse auto, replace
. reg mpg i.rep78 if rep78 == 2, nocons
. predict xb
. tab xb
     Fitted |
     values |      Freq.     Percent        Cum.
------------+-----------------------------------
          0 |         61       88.41       88.41
     19.125 |          8       11.59      100.00
------------+-----------------------------------
      Total |         69      100.00
```

**Comment by jemather** [comment]
Hi Kyle - 

I hit this from the other side while working on a first stage that absorbs the fixed effects instead of materializing them (and I believe will increase speed substantially), where the unidentified levels have to be handled explicitly rather than quietly becoming zeros. The screen I ended up with, in this package's idiom, is:

```stata
tempvar anyuntreated
foreach fe of varlist `absorb' {
    cap drop `anyuntreated'
    qui egen byte `anyuntreated' = max(`touse' & `treatment' == 0), by(`fe')
    qui replace `touse' = 0 if `anyuntreated' == 0
}
```

with a count of the dropped rows reported to the user, and an opt-out for anyone who wants the current behavior.

One pass is enough, which I think isn't obvious: a flagged level has no untreated observations by definition, so every row it drops is a treated row, and dropping treated rows can't reduce any other level's untreated support. So the screen can't cascade and doesn't need to iterate.

With weights the condition should arguably be positive untreated *weight* rather than any untreated row, so that zero-weight untreated observations don't count as support.

Is drop-and-report the semantics you'd want, or would you rather it error and make the user decide?

If you'd want this for the dummy path too, I'm happy to send a PR, with one caveat about scope. `fvrevar` is evaluated on the untreated subsample:

```stata
fvrevar `first_stage' if (`touse' == 1) & (`treatment' == 0)
```

so a level with no untreated observations never gets a column at all. Its rows aren't *omitted* in the `_ms_omit_info` sense — they're all-zero across their block, which is precisely why `predict` hands back 0 instead of missing. Detection therefore can't work off the omission flags; it has to either screen the raw fixed-effect variables before `fvrevar` runs, or identify which generated columns belong to which term. The first is straightforward but only covers plain `i.var` terms in `first_stage()`, leaving interactions like `i.state#c.year` unscreened. I'd rather be explicit that a first pass would have that limitation than ship something that silently misses cases.

**Comment by kylebutts** [comment]
> while working on a first stage that absorbs the fixed effects instead of materializing them (and I believe will increase speed substantially)

Yes, that's what the R version does and is way faster. This would help the first-stage be sped way up: https://github.com/sergiocorreia/reghdfe/issues/17#issuecomment-103703034. The problem is that for the `vcov` calculation, the fixed effects matrix needs to be made and solved with. I do that with sparse matrices in R for the big speed gains. I don't think there's a good equivalent in stata.

Maybe something can be done following the recommendations in here (maybe section 1 option 3)  https://scorreia.com/help/reghdfe_programming.html
